pkg/controller/instance/delta/delta.go

```diff
@@ -180,3 +190,143 @@ func walkSlice(desired, observed []interface{}, path string, differences *[]Diff
 		walkCompare(desired[i], observed[i], newPath, differences)
 	}
 }
+
+// areKubernetesEquivalent checks for Kubernetes-specific semantic equivalence between values.
+// This handles common cases where different representations are semantically equivalent:
+// - Resource quantities: "100m" == "0.1" for CPU
+// - Empty vs nil: [] == null, {} == null
+// - Zero vs nil: 0 == null for optional integer fields
+func areKubernetesEquivalent(desired, observed interface{}, path string) bool {
+	// Direct equality check first
+	if reflect.DeepEqual(desired, observed) {
+		return true
+	}
+
+	// Handle resource quantities (CPU, memory)
+	if isResourceQuantityPath(path) {
+		return areResourceQuantitiesEqual(desired, observed)
+	}
+
+	// Handle empty vs nil arrays
+	if isEmptyVsNilArray(desired, observed) {
+		return true
+	}
+
+	// Handle zero vs nil for optional integer fields
+	if isZeroVsNilInteger(desired, observed) {
+		return true
+	}
+
+	return false
+}
+
+// isResourceQuantityPath checks if the path refers to a Kubernetes resource quantity field
+func isResourceQuantityPath(path string) bool {
```

Comment:
I really think instead of comparing by path (which we would have to maintain for every resource), this should be based on the schema which has this knowledge implicitly no?